qutebrowser/javascript/greasemonkey_wrapper.js

```diff
@@ -148,9 +148,65 @@
         }
     };
 
+    /* TamperMonkey allows assinging to `window` to change the visible global
+     * scope, without breaking other scripts. Eg if the page has set
+     * window.$ for an object for some reason (hello 4chan)
+     * - typeof $ === 'object'
+     * - typeof window.$ == 'undefined'
+     * - window.$ = function() {}
+     * - typeof $ === 'function'
+     * - typeof window.$ == 'function'
+     * Just shadowing `window` won't work because if you try to use '$'
+     * from the global scope you will still get the pages one.
+     * Additionally the userscript expects `window` to actually look
+     * like a fully featured browser window object.
+     *
+     * So let's try to use a Proxy on window and the possibly deprecated
+     * `with` function to make that proxy shadow the global scope.
+     * unsafeWindow should still be the actual global page window.
+     *
+     * There are other Proxy functions that we may need to override.
+     * set, get and has are definitely required.
+     */
     const unsafeWindow = window;
-
-    // ====== The actual user script source ====== //
+    const qute_gm_window_shadow = {};  // stores local changes to window
+    const qute_gm_windowProxyHandler = {
+      get: function(obj, prop) {
+        if (prop in qute_gm_window_shadow)
+          return qute_gm_window_shadow[prop];
+        if (prop in obj) {
+          if (typeof obj[prop] === 'function' && typeof obj[prop].prototype == 'undefined')
+            // Getting TypeError: Illegal Execution when callers try to execute
+            // eg addEventListener from here because they were returned
+            // unbound
+            return obj[prop].bind(obj);
+          return obj[prop];
+        }
+      },
+      set: function(target, prop, val) {
+        return qute_gm_window_shadow[prop] = val;
+      },
+      has: function(target, key) {
+        return key in qute_gm_window_shadow ? true :
+               key in target;
+      }
+    };
+    const qute_gm_window_proxy = new Proxy(unsafeWindow, qute_gm_windowProxyHandler);
+
+    with (qute_gm_window_proxy) {
+      // We can't return `this` or `qute_gm_window_proxy` from
+      // `qute_gm_window_proxy.get('window')` because the Proxy implementation
+      // does typechecking on read-only things. So we have to shadow `window`
+      // more conventionally here. Except we can't do it directly within
+      // with `with` scope because then it would get assigned to the
+      // proxy and we would get the same problem, so we have to make yet
+      // another nested scope.
+      function qute_gm_window_scope() {  // why can't this be anonymous?
```

Comment:
This will not work as expected. Take for example this userscript:

```js
var a = 3;
console.log(a, window.a);
```

It logs `3 undefined`, because the `var` declaration is now in function scope, so the variable shall be local to the function. If above code was run in an actual global scope it would print `3 3`.

Even removing the `qute_gm_window_scope` will not work, because `with` is weird, and variable declarations inside the `with` will only forward to the object passed to `with` if the property already exists on the object, and function declarations aren't forwarded whatsoever.

BTW, this function can not be anonymous because function declarations can't be anonymous (how would you call the function afterwards?). You could replace this with

```js
(function() {
    let window = qute_gm_window_proxy;
    // ====== The actual user script source ====== //
    {{ scriptSource }}
    // ====== End User Script ====== //
})();
```

Reply:
>It logs `3 undefined`

That isn't unexpected and is exactly what tampermonkey logs, which is what I am copying off. I don't think greasemonkey scripts expect to be in global scope, they just expect the things they write to `window` to appear to be in global scope.

Reply:
Aha ok then it was just me expecting a global scope. In that case I can't see why the code shouldn't work.